File: chat-workbench-main/ui/lib/store/message/registry.ts
// Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
// Terms and the SOW between the parties dated 2025.

/**
 * registry.ts
 *
 * This module provides a global registry to track WebSocket handler initialization
 * outside the Zustand store, breaking circular dependencies and preventing
 * multiple handler registrations.
 */

// Track whether WebSocket handlers have been initialized
let wsHandlersInitialized = false;
let wsHandlerInitCount = 0;
let storeCreationCount = 0;
let authLogoutListenerCount = 0;

// Store instance registry
let messageStoreInstance: any = null;

// Debug info for diagnostics
const initTimeStamps: number[] = [];
const initStackTraces: string[] = [];

// Get stack trace for debugging
function getStackTrace(): string {
  const stack = new Error().stack || '';
  return stack.split('\n').slice(2).join('\n');
}

// Get the current timestamp
function timestamp(): string {
  return new Date().toISOString();
}

// API for tracking WebSocket handler initialization
export const wsHandlerRegistry = {
  // Check if handlers are already initialized
  isInitialized: () => wsHandlersInitialized,

  // Mark handlers as initialized
  markInitialized: () => {
    wsHandlersInitialized = true;
    wsHandlerInitCount++;
    initTimeStamps.push(Date.now());
    initStackTraces.push(getStackTrace());
    console.log(
      `[${timestamp()}] WebSocket handlers marked as initialized (count: ${wsHandlerInitCount})`,
    );
  },

  // Get initialization count (for debugging)
  getInitCount: () => wsHandlerInitCount,

  // Get debug info
  getDebugInfo: () => ({
    wsHandlersInitialized,
    wsHandlerInitCount,
    storeCreationCount,
    authLogoutListenerCount,
    initTimeStamps,
    initStackTraces,
  }),
};

// API for tracking store creation
export const storeRegistry = {
  // Check if store is already created
  hasStoreInstance: () => messageStoreInstance !== null,

  // Get store instance
  getStoreInstance: () => messageStoreInstance,

  // Register store instance
  registerStoreInstance: (instance: any) => {
    messageStoreInstance = instance;
    storeCreationCount++;
    console.log(
      `[${timestamp()}] Store instance registered (count: ${storeCreationCount})`,
    );
    return instance;
  },

  // Get store creation count (for debugging)
  getStoreCreationCount: () => storeCreationCount,
};

// API for tracking event listeners
export const listenerRegistry = {
  // Register auth logout listener
  registerAuthLogoutListener: () => {
    authLogoutListenerCount++;
    console.log(
      `[${timestamp()}] Auth logout listener registered (count: ${authLogoutListenerCount})`,
    );
    return authLogoutListenerCount;
  },

  // Get auth logout listener count (for debugging)
  getAuthLogoutListenerCount: () => authLogoutListenerCount,
};

// Log registry usage
console.log(`[${timestamp()}] Registry module loaded`);

File: chat-workbench-main/backend/app/api/middleware/registry.py
# Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
# Terms and the SOW between the parties dated 2025.

"""Middleware registry for managing middleware instances."""

from starlette.middleware.base import BaseHTTPMiddleware


class MiddlewareRegistry:
    """Registry for middleware instances.

    This registry allows middleware instances to be registered and retrieved
    by name, making it easier to configure middleware after initialization.
    """

    def __init__(self) -> None:
        """Initialize the middleware registry."""
        self._middleware: dict[str, BaseHTTPMiddleware] = {}

    def register(self, name: str, middleware: BaseHTTPMiddleware) -> None:
        """Register a middleware instance.

        Args:
            name: The name to register the middleware under
            middleware: The middleware instance to register
        """
        self._middleware[name] = middleware

    def get(self, name: str) -> BaseHTTPMiddleware | None:
        """Get a middleware instance by name.

        Args:
            name: The name of the middleware to retrieve

        Returns:
            The middleware instance, or None if not found
        """
        return self._middleware.get(name)
